### PR TITLE
SO-4447: Move RF2 export configuration from snowowl.yml to CodeSystem properties

### PR DESCRIPTION
**Pull request by molnarlaura**

### CODE REVIEW

File: snomed/com.b2international.snowowl.snomed.core.rest.tests/src/com/b2international/snowowl/snomed/core/rest/io/SnomedExportApiTest.java [hunk 1]
```diff
@@ -1230,6 +1232,55 @@ public void exportUnpublishedOWLExpressionRefsetMembers() throws Exception {
...
+	public void exportRf2WithConfiguration() throws Exception {
+		
+		final Map<String, Object> exportSettings = ImmutableMap.<String, Object>builder()
+				.put("maintainerType",  Rf2MaintainerType.NRC)
+				.put("nrcCountryCode", "GB")
+				.put("extensionNamespaceId", "370137002")
+				.put("refSetLayout", Rf2RefSetExportLayout.COMBINED)
+				.build();
+		
+		String codeSystemId = "CodeSystem-WithOutSettings";
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.core.rest.tests/src/com/b2international/snowowl/snomed/core/rest/io/SnomedExportApiTest.java]: Incorrect CodeSystemId, it should be `SNOMEDCT-custom-rf2-export-config`.

File: snomed/com.b2international.snowowl.snomed.core.rest.tests/src/com/b2international/snowowl/snomed/core/rest/io/SnomedExportApiTest.java [hunk 2]
```diff
@@ -1230,6 +1232,55 @@ public void exportUnpublishedOWLExpressionRefsetMembers() throws Exception {
...
+				.build();
+			
+		assertArchiveContainsFiles(exportArchive, files);
+		
+	}
+	
+	@Test
+	public void exportRf2WithoutConfiguration() throws Exception {
+		
+		String codeSystemId = "CodeSystem-WithSettings";
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.core.rest.tests/src/com/b2international/snowowl/snomed/core/rest/io/SnomedExportApiTest.java]: And here: `Incorrect CodeSystemId, it should be SNOMEDCT-default-rf2-export-config.`

File: snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/SnomedRf2ExportRestService.java [hunk 3]
```diff
@@ -79,7 +80,7 @@
 			.setIncludePreReleaseContent(params.isIncludeUnpublished())
 			.setModules(params.getModuleIds())
 			.setRefSets(params.getRefSetIds())
-			.setCountryNamespaceElement(params.getNamespaceId())
+			.setCountryNamespaceElement(getCountryNamespaceElement(params))
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/SnomedRf2ExportRestService.java]: What happened with the `namespaceId` argument? It's no longer being used, which is probably not what we want.

**Review by cmark — CHANGES_REQUESTED** [review]

File: snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedRf2ExportConfiguration.java [hunk 4]
```diff
@@ -57,11 +56,8 @@
 	@ApiParam(value = "The RF2 RefSet file layout to use. Defaults to server configuration key 'snomed.export.refSetLayout'.", allowableValues = "combined,individual")
 	private String refSetLayout;
 	
-	@ApiParam(value = "The maintainerType to use un the release archive name")
-	private Rf2MaintainerType maintainerType = Rf2MaintainerType.SNOMED_INTERNATIONAL;
-	
-	@ApiParam(value = "The nrcCountryCode to use un the release archive name")
-	private String nrcCountryCode = "";
+	@ApiParam(value = "The CodeSystem's id on wich the export was initiated.")
+	private String codeSystemId;
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedRf2ExportConfiguration.java]: But this is already specified via the path parameter, why do you need this? Also, the CodeSystem can be calculated from the current branch path on the fly, so feel free to remove this.

File: snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedRf2ExportConfiguration.java [hunk 5]
```diff
@@ -57,11 +56,8 @@
 	@ApiParam(value = "The RF2 RefSet file layout to use. Defaults to server configuration key 'snomed.export.refSetLayout'.", allowableValues = "combined,individual")
 	private String refSetLayout;
 	
-	@ApiParam(value = "The maintainerType to use un the release archive name")
-	private Rf2MaintainerType maintainerType = Rf2MaintainerType.SNOMED_INTERNATIONAL;
-	
-	@ApiParam(value = "The nrcCountryCode to use un the release archive name")
-	private String nrcCountryCode = "";
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedRf2ExportConfiguration.java]: Why did you remove these?

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java [hunk 6]
```diff
@@ -216,8 +218,73 @@ void setExtensionOnly(final boolean extensionOnly) {
...
+				break;
+		}
+		
+		return builder.toString();
+	}
+	
 	@Override
 	public Attachment execute(final BranchContext context) {
+		
+		SnomedExportDefaultConfiguration exportDefaultConfiguration = ApplicationContext.getServiceForClass(SnomedCoreConfiguration.class).getExport();
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java]: Please remove the `snomed.export.*` configuration keys, you are replacing these with the ones in the CodeSystem settings.

**Review by cmark — CHANGES_REQUESTED** [review]

File: snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedRf2ExportConfiguration.java (hunk 5 above)
Comment (molnarlaura) [reply snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedRf2ExportConfiguration.java]: I restored them.

File: snomed/com.b2international.snowowl.snomed.core.rest.tests/src/com/b2international/snowowl/snomed/core/rest/io/SnomedExportApiTest.java [hunk 7]
```diff
@@ -1230,6 +1232,59 @@ public void exportUnpublishedOWLExpressionRefsetMembers() throws Exception {
...
+		final Map<String, Boolean> files = ImmutableMap.<String, Boolean>builder()
+				.put("der2_sssssssRefset_MRCMDomainSnapshot_INT", true)
+				.put("sct2_Relationship_Snapshot_INT", true)
+				.put("sct2_Description_Snapshot-en_INT", true)
+				.put("der2_sRefset_SimpleMapSnapshot_INT", true)
+				.build();
+			
+		assertArchiveContainsFiles(exportArchive, files);
+		
+	}
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.core.rest.tests/src/com/b2international/snowowl/snomed/core/rest/io/SnomedExportApiTest.java:1415]: It would be nice to have additional tests that verify different export configuration, like defining both namespace, maintainerType and other properties and asserting the end file name got computed properly or not.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java [hunk 8]
```diff
@@ -226,6 +257,39 @@ public Attachment execute(final BranchContext context) {
 		// Step 1: check if the export reference branch is a working branch path descendant
 		final TerminologyResource referenceCodeSystem = context.service(TerminologyResource.class);
 
+		if (referenceCodeSystem.getSettings() != null && !referenceCodeSystem.getSettings().isEmpty()) {
+			if (Strings.isNullOrEmpty(countryNamespaceElement)) {
+
+				if (maintainerType == null) {
+
+					String maintainerType = (String) referenceCodeSystem.getSettings().get("maintainerType");
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java]: These config keys should be moved next to the currently available config keys. See `SnomedTerminologyComponentConstants`.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java [hunk 9]
```diff
@@ -226,6 +257,39 @@ public Attachment execute(final BranchContext context) {
...
+					if(!Strings.isNullOrEmpty(maintainerType)) {
+						String customCountryNamespaceElement = getCountryNamespaceElement(Rf2MaintainerType.getByNameIgnoreCase(maintainerType), Strings.nullToEmpty(nrcCountryCode));
+						countryNamespaceElement = customCountryNamespaceElement;
+					}
+				} else {
+					countryNamespaceElement = getCountryNamespaceElement(maintainerType, Strings.nullToEmpty(nrcCountryCode));
+				}
+			}
+
+			if (refSetExportLayout == null && referenceCodeSystem.getSettings().get("refSetLayout") != null) {
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java]: `containsKey` is must better that a `get` and `null` check.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java [hunk 10]
```diff
@@ -226,6 +257,39 @@ public Attachment execute(final BranchContext context) {
 		// Step 1: check if the export reference branch is a working branch path descendant
 		final TerminologyResource referenceCodeSystem = context.service(TerminologyResource.class);
 
+		if (referenceCodeSystem.getSettings() != null && !referenceCodeSystem.getSettings().isEmpty()) {
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java]: Use `CompareUtils.isEmpty()` for null + emptiness checks.

**Review by cmark — CHANGES_REQUESTED** [review]

**Review by cmark — APPROVED** [review]
LGTM!
